stubs/composer.stub

```
@@ -0,0 +1,11 @@
+<?php
+
+namespace Composer\Package;
+
+/**
+ * @phpstan-type AutoloadRules array{psr-0?: array<string, string[]>, psr-4?: array<string, string[]>, classmap?: list<string>, files?: list<string>, exclude-from-classmap?: list<string>}
+ * @phpstan-type DevAutoloadRules array{psr-0?: array<string, string[]>, psr-4?: array<string, string[]>, classmap?: list<string>, files?: list<string>}
+ */
+interface PackageInterface
+{
+}
```

how did you generate this file?

---

Manually. It’s just to fix a PHPdoc type that’s incorrect in Composer’s source code and make PHPStan happy.

---

Basically if you look at the original PackageInterface you’ll see the PHPDoc type is nearly the same, I think I just removed an `|null` or similar.

---

couldn't we have a 1 line baseline file? just because this would be something for a PR to composer and what happens when composer changes? with a baseline we'll get an error and we'll be able to remove the baseline completely

---

A stub is better than baselining the error, because the ignored error could suppress other legitimate errors.

Since it was only correcting one type, I removed the stub file and just directly set it in `AutoloadPlugin.php`

---

with phpstan2 you can baseline the single type of error, not even the whole line ;-)

---

But it has downstream effects. Without the fixed docblock if we check:

```php
\PHPStan\dumpType($autoloadDefinition['psr-0']['M']); # Dumped type: array<string>|string
```

It reports the PSR-0 prefixes we built in `AutoloadRuntime.php` could be a string, but that's incorrect because it's always an array.

Later if we decided to loop through `$autoloadDefinition` we'd have to be checking if it's a string or array, or just baseline more errors.

---

no no i didn't want to convince you for the baseline, it was just a comment about phpstan2. also, you know that phpstan itself runs on level 8 and Ondrej himself told he doesn't want to raise it heheheheh

---

I misunderstood. 😃

This particular error happens at level 7, but anyway getting to level 10 was really no problem since this repo is nowhere near the size of PHPStan itself. Plus composer has it's own set of rules specifically for extensions so it's nice to check those too.